File: client/client-multi/client-java-core/src/main/java/org/kaaproject/kaa/client/event/DefaultEventManager.java
/*
 * Copyright 2014-2016 CyberVision, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.kaaproject.kaa.client.event;

import org.kaaproject.kaa.client.channel.EventTransport;
import org.kaaproject.kaa.client.context.ExecutorContext;
import org.kaaproject.kaa.client.persistence.KaaClientState;
import org.kaaproject.kaa.client.transact.TransactionId;
import org.kaaproject.kaa.common.endpoint.gen.Event;
import org.kaaproject.kaa.common.endpoint.gen.EventListenersRequest;
import org.kaaproject.kaa.common.endpoint.gen.EventListenersResponse;
import org.kaaproject.kaa.common.endpoint.gen.EventSyncRequest;
import org.kaaproject.kaa.common.endpoint.gen.SyncResponseResultType;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.nio.ByteBuffer;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.HashSet;
import java.util.LinkedList;
import java.util.List;
import java.util.Map;
import java.util.Random;
import java.util.Set;

/**
 * Default {@link EventManager} implementation.
 *
 * @author Taras Lemkin
 */
public class DefaultEventManager implements EventManager {

  private static final Logger LOG = LoggerFactory.getLogger(DefaultEventManager.class);
  private final Set<EventFamily> registeredEventFamilies = new HashSet<EventFamily>();
  private final List<Event> currentEvents = new LinkedList<Event>();
  private final Object eventsGuard = new Object();
  private final Object trxGuard = new Object();
  private final Map<Integer, EventListenersRequestBinding> eventListenersRequests =
          new HashMap<Integer, EventListenersRequestBinding>();
  private final EventTransport transport;
  private final KaaClientState state;
  private final ExecutorContext executorContext;
  private final Map<TransactionId, List<Event>> transactions = new HashMap<>();
  private Boolean isEngaged = false;

  /**
   * All-args constructor.
   */
  public DefaultEventManager(KaaClientState state, ExecutorContext executorContext,
                             EventTransport transport) {
    this.state = state;
    this.transport = transport;
    this.executorContext = executorContext;
  }

  @Override
  public void fillEventListenersSyncRequest(EventSyncRequest request) {
    if (!eventListenersRequests.isEmpty()) {
      LOG.debug("There are {} unresolved eventListenersResolution request{}",
              eventListenersRequests.size(),
          (eventListenersRequests.size() == 1 ? "" : "s")); // NOSONAR
      List<EventListenersRequest> requests = new ArrayList<EventListenersRequest>();
      for (Map.Entry<Integer, EventListenersRequestBinding> entry : eventListenersRequests
              .entrySet()) {
        if (!entry.getValue().isSent()) {
          requests.add(entry.getValue().getRequest());
          entry.getValue().setSent(Boolean.TRUE);
        }
      }
      request.setEventListenersRequests(requests);
    }
  }

  @Override
  public void clearState() {
    synchronized (eventsGuard) {
      currentEvents.clear();
    }
  }

  @Override
  public void produceEvent(String eventFqn, byte[] data, String target) {
    produceEvent(eventFqn, data, target, null);
  }

  @Override
  public void produceEvent(String eventFqn, byte[] data, String target, TransactionId trxId) {
    if (trxId == null) {
      LOG.info("Producing event [eventClassFQN: {}, target: {}]", eventFqn,
              (target != null ? target : "broadcast")); // NOSONAR
      synchronized (eventsGuard) {
        currentEvents.add(new Event(state.getAndIncrementEventSeqNum(), eventFqn,
                ByteBuffer.wrap(data), null, target));
      }
      if (!isEngaged) {
        transport.sync();
      }
    } else {
      LOG.info("Adding event [eventClassFQN: {}, target: {}] to transaction {}",
              eventFqn, target != null ? target : "broadcast", trxId); // NOSONAR
      synchronized (trxGuard) {
        List<Event> events = transactions.get(trxId);
        if (events != null) {
          events.add(new Event(-1, eventFqn, ByteBuffer.wrap(data), null, target));
        } else {
          LOG.warn("Transaction with id {} is missing. Ignoring event");
        }
      }
    }
  }

  @Override
  public void registerEventFamily(EventFamily eventFamily) {
    registeredEventFamilies.add(eventFamily);
  }

  @Override
  public void onGenericEvent(final String eventFqn, final byte[] data, final String source) {
    LOG.info("Received event [eventClassFQN: {}]", eventFqn);
    for (final EventFamily family : registeredEventFamilies) {
      LOG.info("Lookup event fqn {} in family {}", eventFqn, family);
      if (family.getSupportedEventFqns().contains(eventFqn)) {
        LOG.info("Event fqn {} found in family {}", eventFqn, family);
        executorContext.getCallbackExecutor().submit(new Runnable() {
          @Override
          public void run() {
            family.onGenericEvent(eventFqn, data, source);
          }
        });
      }
    }
  }

  @Override
  public int findEventListeners(List<String> eventFqns, FindEventListenersCallback listener) {
    int requestId = new Random().nextInt();
    EventListenersRequest request = new EventListenersRequest(requestId, eventFqns);
    EventListenersRequestBinding bind = new EventListenersRequestBinding(listener, request);
    eventListenersRequests.put(requestId, bind);
    LOG.debug("Adding event listener resolution request. Request ID: {}", requestId);
    if (!isEngaged) {
      transport.sync();
    }
    return requestId;
  }

  @Override
  public void eventListenersResponseReceived(List<EventListenersResponse> response) {
    for (final EventListenersResponse singleResponse : response) {
      LOG.debug("Received event listener resolution response: {}", response);
      final EventListenersRequestBinding bind = eventListenersRequests.remove(
              singleResponse.getRequestId());
      executorContext.getCallbackExecutor().submit(new Runnable() {
        @Override
        public void run() {
          if (bind != null) {
            if (singleResponse.getResult() == SyncResponseResultType.SUCCESS) {
              bind.getListener().onEventListenersReceived(singleResponse.getListeners());
            } else {
              bind.getListener().onRequestFailed();
            }
          }
        }
      });
    }
  }

  @Override
  public List<Event> pollPendingEvents() {
    return getPendingEvents(true);
  }

  @Override
  public List<Event> peekPendingEvents() {
    return getPendingEvents(false);
  }

  private List<Event> getPendingEvents(boolean clear) {
    List<Event> pendingEvents = new ArrayList<Event>();
    synchronized (eventsGuard) {
      pendingEvents.addAll(currentEvents);
      if (clear) {
        currentEvents.clear();
      }
    }
    return pendingEvents;
  }

  @Override
  public TransactionId beginTransaction() {
    TransactionId trxId = new TransactionId();
    synchronized (trxGuard) {
      if (!transactions.containsKey(trxId)) {
        LOG.debug("Creating events transaction with id {}", trxId);
        transactions.put(trxId, new LinkedList<Event>());
      }
    }
    return trxId;
  }

  @Override
  public void commit(TransactionId trxId) {
    LOG.debug("Commiting events transaction with id {}", trxId);
    synchronized (trxGuard) {
      List<Event> eventsToCommit = transactions.remove(trxId);
      synchronized (eventsGuard) {
        for (Event e : eventsToCommit) {
          e.setSeqNum(state.getAndIncrementEventSeqNum());
          currentEvents.add(e);
        }
      }
      if (!isEngaged) {
        transport.sync();
      }

    }
  }

  @Override
  public void rollback(TransactionId trxId) {
    LOG.debug("Rolling back events transaction with id {}", trxId);
    synchronized (trxGuard) {
      List<Event> eventsToRemove = transactions.remove(trxId);
      if (eventsToRemove != null) {
        for (Event e : eventsToRemove) {
          LOG.trace("Removing event {}", e);
        }
      } else {
        LOG.debug("Transaction with id {} was not created", trxId);
      }
    }
  }

  @Override
  public synchronized void engageDataChannel() {
    isEngaged = true;
  }

  @Override
  public synchronized boolean releaseDataChannel() {
    isEngaged = false;
    boolean needSync = !currentEvents.isEmpty();
    if (!needSync) {
      for (EventListenersRequestBinding b : eventListenersRequests.values()) {
        needSync |= !b.isSent();
      }
    }
    return needSync;
  }

  private class EventListenersRequestBinding {
    private final FindEventListenersCallback listener;
    private final EventListenersRequest request;
    private Boolean sent;

    public EventListenersRequestBinding(
            FindEventListenersCallback listener, EventListenersRequest request) {
      this.listener = listener;
      this.request = request;
      this.sent = false;
    }

    public FindEventListenersCallback getListener() {
      return listener;
    }

    public EventListenersRequest getRequest() {
      return request;
    }

    public Boolean isSent() {
      return sent;
    }

    public void setSent(Boolean sent) {
      this.sent = sent;
    }
  }

}

File: client/client-multi/client-objective-c/Kaa/event/DefaultEventManager.m
/*
 * Copyright 2014-2016 CyberVision, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#import "DefaultEventManager.h"
#import "EventListenersRequestBinding.h"
#import "EndpointGen.h"
#include <stdlib.h>
#import "KaaLogging.h"

#define TAG @"DefaultEventManager >>>"

@interface DefaultEventManager ()

@property (nonatomic, strong) id<KaaClientState> state;;
@property (nonatomic, strong) id<ExecutorContext> executorContext;
@property (nonatomic, strong) id<EventTransport> transport;

@property (nonatomic, strong) NSMutableSet *registeredEventFamilies;         //<BaseEventFamily>
@property (nonatomic, strong) NSMutableArray *currentEvents;                 //<Event>
@property (nonatomic, strong) NSMutableDictionary *eventListenersRequests;   //<NSNumber, EventListenersRequestBinding>
@property (nonatomic, strong) NSMutableDictionary *transactions;             //<TransactionId, NSArray<Event>>

@property (nonatomic) BOOL isEngaged;
@property (nonatomic) int requestId;

@property (nonatomic, strong) NSObject *eventGuard;
@property (nonatomic, strong) NSObject *transactionsGuard;

- (NSMutableArray *)getAndClearPendingEvents:(BOOL)clear;

@end

@implementation DefaultEventManager

- (instancetype)initWithState:(id<KaaClientState>)state
              executorContext:(id<ExecutorContext>)executorContext
               eventTransport:(id<EventTransport>)transport {
    self = [super init];
    if (self) {
        self.state = state;
        self.executorContext = executorContext;
        self.transport = transport;
        
        self.registeredEventFamilies = [NSMutableSet set];
        self.currentEvents = [NSMutableArray array];
        self.eventListenersRequests = [NSMutableDictionary dictionary];
        self.transactions = [NSMutableDictionary dictionary];
        
        self.requestId = 0;
        self.isEngaged = NO;
        self.eventGuard = [[NSObject alloc] init];
        self.transactionsGuard = [[NSObject alloc] init];
    }
    return self;
}

- (void)fillEventListenersSyncRequest:(EventSyncRequest *)request {
    if ([self.eventListenersRequests count] > 0) {
        DDLogDebug(@"%@ Unresolved eventListenersResolution request count: %li",
                   TAG, (long)[self.eventListenersRequests count]);
        NSMutableArray *requests = [NSMutableArray array];
        for (EventListenersRequestBinding *bind in self.eventListenersRequests.allValues) {
            if (!bind.isSent) {
                [requests addObject:bind.request];
                bind.isSent = YES;
            }
        }
        request.eventListenersRequests = [KAAUnion unionWithBranch:KAA_UNION_ARRAY_EVENT_LISTENERS_REQUEST_OR_NULL_BRANCH_0 data:requests];
    }
}

- (void)clearState {
    @synchronized(self.eventGuard) {
        [self.currentEvents removeAllObjects];
    }
}

- (void)produceEventWithFQN:(NSString *)eventFQN data:(NSData *)data target:(NSString *)target {
    [self produceEventWithFQN:eventFQN data:data target:target transactionId:nil];
}

- (void)produceEventWithFQN:(NSString *)eventFQN data:(NSData *)data target:(NSString *)target transactionId:(TransactionId *)transactionId {
    if (transactionId) {
        DDLogInfo(@"%@ Adding event [eventClassFQN: %@, target: %@] to transaction %@", TAG, eventFQN, (target ?: @"broadcast"), transactionId);
        @synchronized(self.transactionsGuard) {
            NSMutableArray *events = self.transactions[transactionId];
            if (events) {
                Event *event = [[Event alloc] init];
                event.seqNum = -1;
                event.eventClassFQN = eventFQN;
                event.eventData = [NSData dataWithData:data];
                if (target) {
                    event.target = [KAAUnion unionWithBranch:KAA_UNION_STRING_OR_NULL_BRANCH_0 data:target];
                }
                [events addObject:event];
            } else {
                DDLogWarn(@"%@ Transaction with id %@ is missing. Ignoring event.", TAG, transactionId);
            }
        }
    } else {
        DDLogInfo(@"%@ Producing event [eventClassFQN: %@, target: %@]", TAG, eventFQN, (target ?: @"broadcast"));
        @synchronized(self.eventGuard) {
            Event *event = [[Event alloc] init];
            event.seqNum = [self.state getAndIncrementEventSequenceNumber];
            event.eventClassFQN = eventFQN;
            event.eventData = [NSData dataWithData:data];
            if (target) {
                event.target = [KAAUnion unionWithBranch:KAA_UNION_STRING_OR_NULL_BRANCH_0 data:target];
            }
            [self.currentEvents addObject:event];
        }
        
        if (!self.isEngaged) {
            [self.transport sync];
        }
    }
}

- (void)registerEventFamily:(id<BaseEventFamily>)eventFamily {
    [self.registeredEventFamilies addObject:eventFamily];
}

- (void)onGenericEvent:(NSString *)eventFQN withData:(NSData *)data fromSource:(NSString *)source {
    DDLogInfo(@"%@ Received event [eventClassFQN: %@]", TAG, eventFQN);
    for (id<BaseEventFamily> family in self.registeredEventFamilies) {
        DDLogInfo(@"%@ Lookup event fqn %@ in family %@", TAG, eventFQN, family);
        if ([[family getSupportedEventFQNs] containsObject:eventFQN]) {
            DDLogInfo(@"%@ Event fqn [%@] found in family [%@]", TAG, eventFQN, family);
            [[self.executorContext getCallbackExecutor] addOperationWithBlock:^{
                [family onGenericEvent:eventFQN withData:data fromSource:source];
            }];
        }
    }
}

- (int32_t)requestListenersForEventFQNs:(NSArray *)eventFQNs delegate:(id<FindEventListenersDelegate>)delegate {
    int32_t requestId = self.requestId++;
    EventListenersRequest *request = [[EventListenersRequest alloc] init];
    request.requestId = requestId;
    request.eventClassFQNs = eventFQNs;
    EventListenersRequestBinding *bind = [[EventListenersRequestBinding alloc] initWithRequest:request delegate:delegate];
    self.eventListenersRequests[@(requestId)] = bind;
    DDLogDebug(@"%@ Adding event listener resolution request. Request ID: %i", TAG, requestId);
    if (!self.isEngaged) {
        [self.transport sync];
    }
    return requestId;
}

- (void)eventListenersResponseReceived:(NSArray *)response {
    for (EventListenersResponse *singleResponse in response) {
        DDLogDebug(@"%@ Received event listener resolution response: %@", TAG, singleResponse);
        __block EventListenersRequestBinding *bind = self.eventListenersRequests[@(singleResponse.requestId)];
        if (bind) {
            [self.eventListenersRequests removeObjectForKey:@(singleResponse.requestId)];
            [[self.executorContext getCallbackExecutor] addOperationWithBlock:^{
                if (singleResponse.result == SYNC_RESPONSE_RESULT_TYPE_SUCCESS) {
                    [bind.delegate onEventListenersReceived:((NSArray *)singleResponse.listeners.data)];
                } else {
                    [bind.delegate onRequestFailed];
                }
            }];
        }
    }
}

- (NSArray *)pollPendingEvents {
    return [self getAndClearPendingEvents:YES];
}

- (NSArray *)peekPendingEvents {
    return [self getAndClearPendingEvents:NO];
}

- (NSMutableArray *)getAndClearPendingEvents:(BOOL)clear {
    @synchronized(self.eventGuard) {
        NSMutableArray *pendingEvents = [NSMutableArray arrayWithArray:self.currentEvents];
        if (clear) {
            [self.currentEvents removeAllObjects];
        }
    return pendingEvents;
    }
}

- (TransactionId *)beginTransaction {
    TransactionId *trxId = [[TransactionId alloc] init];
    @synchronized(self.transactionsGuard) {
        if (!self.transactions[trxId]) {
            DDLogDebug(@"%@ Creating events transaction with id [%@]", TAG, trxId);
            self.transactions[trxId] = [NSMutableArray array];
        }
    }
    return trxId;
}

- (void)commitTransactionWithId:(TransactionId *)trxId {
    DDLogDebug(@"%@ Committing events transaction with id [%@]", TAG, trxId);
    @synchronized(self.transactionsGuard) {
        NSArray *eventsToCommit = self.transactions[trxId];
        if (eventsToCommit) {
            [self.transactions removeObjectForKey:trxId];
            
            @synchronized(self.eventGuard) {
                for (Event *event in eventsToCommit) {
                    event.seqNum = [self.state getAndIncrementEventSequenceNumber];
                    [self.currentEvents addObject:event];
                }
            }
        }
        if (!self.isEngaged) {
            [self.transport sync];
        }
    }
}

- (void)rollbackTransactionWithId:(TransactionId *)trxId {
    DDLogDebug(@"%@ Rolling back events transaction with id %@", TAG, trxId);
    @synchronized(self.transactionsGuard) {
        NSMutableArray *eventsToRemove = self.transactions[trxId];
        if (eventsToRemove) {
            [self.transactions removeObjectForKey:trxId];
            for (Event *event in eventsToRemove) {
                DDLogVerbose(@"%@ Removing event %@", TAG, event);
            }
        } else {
            DDLogDebug(@"%@ Transaction with id [%@] was not created", TAG, trxId);
        }
    }
}

- (void)engageDataChannel {
    @synchronized (self) {
        self.isEngaged = YES;
    }
}

- (BOOL)releaseDataChannel {
    @synchronized (self) {
        self.isEngaged = NO;
        BOOL needSync = [self.currentEvents count] > 0;
        if (!needSync) {
            for (EventListenersRequestBinding *bind in self.eventListenersRequests.allValues) {
                needSync |= !bind.isSent;
            }
        }
        return needSync;
    }
}

@end
